### PR TITLE
Particle system - Part3

### PR DESCRIPTION
**Pull request by caguero**
It depends on https://github.com/ignitionrobotics/ign-msgs/pull/127 .

This pull request connects the particle emitter system with the rendering engine and we should start visualizing the particles.

You can test it with: `ign gazebo -r -v 4 ../examples/worlds/particle_emitter.sdf`

gazebo_particles

### CODE REVIEW

**Comment by iche033** [comment]
nice! can you add `particle_emitter.sdf` world?

**Comment by caguero** [comment]
> nice! can you add `particle_emitter.sdf` world?

Added c76b26b :)

File: include/ignition/gazebo/rendering/SceneManager.hh [hunk 1]
```diff
@@ -36,6 +36,7 @@
 
 #include <ignition/rendering/RenderTypes.hh>
 
+#include <ignition/gazebo/components/ParticleEmitter.hh>
```
Comment (adlarkin) [inline include/ignition/gazebo/rendering/SceneManager.hh:39]: I don't see the `ParticleEmitter` component being used in this file, but I do see `msgs::ParticleEmitter` being used. Should we actually be including `ignition/msgs/particle_emitter.pb.h` instead?

File: src/rendering/SceneManager.cc [hunk 2]
```diff
@@ -74,6 +78,10 @@ class ignition::gazebo::SceneManagerPrivate
   /// \brief Map of light entity in Gazebo to light pointers.
   public: std::map<Entity, rendering::LightPtr> lights;
 
+  /// \brief Map of particle emitter entity in Gazebo to particle emitter
+  /// pointers.
```
Comment (adlarkin) [inline src/rendering/SceneManager.cc:82]:
```suggestion
  /// rendering pointers.
```

File: src/rendering/SceneManager.cc [hunk 3]
```diff
@@ -929,6 +937,117 @@ rendering::LightPtr SceneManager::CreateLight(Entity _id,
...
+  std::string name = _emitter.name().empty() ? std::to_string(_id) :
+    _emitter.name();
+  if (parent)
+    name = parent->Name() +  "::" + name;
+
+  rendering::ParticleEmitterPtr emitter;
+  emitter = this->dataPtr->scene->CreateParticleEmitter(name);
+
+  // Type.
+  switch (_emitter.type())
```
Comment (adlarkin) [inline src/rendering/SceneManager.cc:978]: Before we run a switch on `_emitter.type()` to set the type for `emitter`, can we include a check to make sure `emitter` isn't `NULL`? 

(https://github.com/ignitionrobotics/ign-rendering/blob/cac86c83d488b1866374013a92e635ce604a1eee/include/ignition/rendering/Scene.hh#L941-L947)

File: src/rendering/SceneManager.cc [hunk 4]
```diff
@@ -929,6 +937,117 @@ rendering::LightPtr SceneManager::CreateLight(Entity _id,
...
+  {
+    case ignition::msgs::ParticleEmitter_EmitterType_BOX:
+    {
```
Comment (adlarkin) [inline src/rendering/SceneManager.cc:981]: Do we usually include `{` and `}` for cases in a switch statement? I've seen other switch statements in `ign-gazebo` leave them out. I'm not sure if this matters or not.

File: src/rendering/SceneManager.cc [hunk 5]
```diff
@@ -990,22 +1111,17 @@ rendering::NodePtr SceneManager::NodeById(Entity _id) const
...
-    }
+    return lIt->second;
+  }
+  auto sIt = this->dataPtr->sensors.find(_id);
+  if (sIt != this->dataPtr->sensors.end())
+  {
+    return sIt->second;
   }
+
   return rendering::NodePtr();
```
Comment (adlarkin) [inline src/rendering/SceneManager.cc:1125]: Should we also add code to `SceneManager::NodeById` that searches for a node in [`this->dataPtr->ParticleEmitters`](https://github.com/ignitionrobotics/ign-gazebo/blob/502abfa1a1d2515a71a66bffe86bdae2a06c7e72/src/rendering/SceneManager.cc#L81-L83)? Looking at the code in `ign-rendering`, [a ParticleEmitter virtually inherits a Visual](https://github.com/ignitionrobotics/ign-rendering/blob/8bb934d394fdbd54cbda81b6fb530b68813bfdd9/include/ignition/rendering/ParticleEmitter.hh#L54-L59), and [a VIsual virtually inherits a node](https://github.com/ignitionrobotics/ign-rendering/blob/cac86c83d488b1866374013a92e635ce604a1eee/include/ignition/rendering/Visual.hh#L34-L38). So, I would think that `this->dataPtr->ParticleEmitters` can be searched for nodes as well.

File: src/rendering/SceneManager.cc (hunk 4 above)
Comment (nkoenig) [reply src/rendering/SceneManager.cc:981]: It's okay to include them. They are mandatory in some situations, such as creating a new variable within the `case`.

File: src/rendering/SceneManager.cc (hunk 5 above)
Comment (nkoenig) [reply src/rendering/SceneManager.cc:1125]: @iche033 , can you address this comment?

File: examples/worlds/particle_emitter.sdf [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            <ambient>0.8 0.8 0.8 1</ambient>
+            <diffuse>0.8 0.8 0.8 1</diffuse>
+            <specular>0.8 0.8 0.8 1</specular>
+          </material>
+        </visual>
+      </link>
+    </model>
+
+    <include>
+      <uri>https://fuel.ignitionrobotics.org/1.0/caguero/models/smoke_generator/2</uri>
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply examples/worlds/particle_emitter.sdf]: Done in defd33f6809b8b204ee91426bd48b6eb4c54fc7e

File: include/ignition/gazebo/rendering/SceneManager.hh (hunk 1 above)
Comment (nkoenig) [reply include/ignition/gazebo/rendering/SceneManager.hh]: Done in defd33f6809b8b204ee91426bd48b6eb4c54fc7e

File: src/rendering/SceneManager.cc (hunk 3 above)
Comment (nkoenig) [reply src/rendering/SceneManager.cc]: Done in defd33f6809b8b204ee91426bd48b6eb4c54fc7e

File: src/rendering/SceneManager.cc (hunk 2 above)
Comment (nkoenig) [reply src/rendering/SceneManager.cc]: Done in 2e28aae0d51a9abe35018e891f6c87731e24a76f

File: src/rendering/SceneManager.cc (hunk 5 above)
Comment (iche033) [reply src/rendering/SceneManager.cc:1136]: yep done. f58b197aa31669e4323a4e5519dab7fb20dc2df5

**Pull request merged**